### ISSUE TITLE
GeospatialCoordinateEncoder should use same calculation as in HTM.java

### ISSUE THREAD

**Issue by cogmission** [numenta/nupic-legacy#1567]
HTM.java's GeospatialCoordinateEncoder uses the direct calculation skimmed from a large library which can do the Lat/Long to Mercator projection and back in 3 to 4 lines of code. (Python would be even shorter). All NuPIC languages should use the same calculation found here:

https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/encoders/GeospatialCoordinateEncoder.java#L75

**Comment by chessweb01** [comment]
I'd like to work on this issue. Could someone assign it to me?

**Comment by breznak** [comment]
@chessweb01 assigned you, thanks for working on it! 
Btw, can someone explain the issue to me? It says Nupic should use a geospatial transformation and then points to nupic code for it's example implementation? CC @cogmission

**Comment by cogmission** [comment]
The link marks the exact position in the geospatial code in HTM.java?

**Comment by breznak** [comment]
sorry, it's java, not python :tongue: I should get myself a coffee!

**Comment by cogmission** [comment]
Give me one too. I take cream and sugar. :-)

**Comment by cogmission** [comment]
… actually honey I don't really use sugar

**Comment by cogmission** [comment]
Also the issue is about replacing the longer code with some short hand code that I found to convert to and from Mercator projection. The calculations would work in any language, so Chetan suggested that we convert the Python code over to those calculations because they are shorter.

**Comment by chessweb01** [comment]
First a couple of observations (please correct any misconception):

1. In Python the calculations in question are performed via calls to two functions: `proj` and `transform`. Those functions are imported from library `pyproj`. The function calls happen in `GeospatialCoordinateEncoder::coordinateForPosition(self, longitude, latitude, altitude=None)`, but `transform` is called only in case there is an altitude different from `None`.

2. In htm.jave there are two methods relevant to the coordinate calculations: `toMercator` and `inverseMercator`. As far as I can see, `inverseMercator` is not used anywhere in htm.java. `toMercator` is called in `GeospatialCoordinateEncoder::coordinateForPosition` and thus corresponds to the Python call to `proj`. However, `toMercator`does not allow for an altitude parameter. Therefore htm.java neither needs nor provides a function corresponding to Python's `transform`.        

Now my question is this:
Is it sufficient to just replace the `proj` call by the `toMercator` calculations in order to resolve this issue or am I missing something here? I'm asking because due to `tranform` nupic would still depend on `pyproj` if one wants to allow for an altitude.

**Comment by chetan51** [comment]
@chessweb01 Ideally, htm.java would also support the altitude parameter, and `toMercator` would support the altitude parameter as well. For now, it should be sufficient to replace the `proj` call by the `toMercator` calculations.

**Comment by cogmission** [comment]
Yes, there was the "bounty" work done to include the "altitude" parameter
that never got back ported into htm.java - we need to look into that...

On Tue, Apr 14, 2015 at 12:35 PM, Chetan Surpur <notifications@github.com>
wrote:

> @chessweb01 <https://github.com/chessweb01> Ideally, htm.java would also
> support the altitude parameter, and toMercator would support the altitude
> parameter as well. For now, it should be sufficient to replace the proj
> call by the toMercator calculations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1567#issuecomment-92991352>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by chessweb01** [comment]
Looking at the history of `geospatial_coordinate.py`, I noticed that the optional altitude parameter that needs `Proj::transform` was added a couple of weeks after @cogmission opened this issue. So I think the proper way to go is this: In a first step do the long/lat calculation without altitude using the direct calculation from htm.java. In a second step, scrape the relevant equations from `Proj::transform` and implement the altitude transformations directly, as well, thus removing the pyproj dependency completely. Any comments?

**Comment by cogmission** [comment]
:+1: Precisely! Then I can port that bad fella!  :-)

**Comment by chessweb01** [comment]
@cogmission:
In the Proj.4 project that is the basis for pyproj I found in `geocent.c` the function `pj_Convert_Geodetic_To_Geocentric`. I simplified it according to epsg:3785 and ported it to Java.  Could you please check if this is what we are after and - if yes - whether my simplifications are correct? I'd hate implementing it into nupic if it is off the mark.

    /**
     * Returns geocentric coordinates converted from Long/Lat/Alt
     * 
     * @param lon	the longitude
     * @param lat	the lattitude
     * @param alt      the altitude in m
     * @return
     */
    protected double[] toGeocentric(double lon, double lat, double alt) {

        lon *= Math.PI / 180;
        lat *= Math.PI / 180;

        double PI_OVER_2 = Math.PI / 2;
	
        if (lat < -PI_OVER_2 && lat > -1.001 * PI_OVER_2 ) {
            lat = -PI_OVER_2;
        } else if ( lat > PI_OVER_2 && lat < 1.001 * PI_OVER_2 ) {
            lat = PI_OVER_2;
        } else if (lat < -PI_OVER_2 || lat > PI_OVER_2) { 
            throw new IllegalArgumentException("latitude out of range");
        }

        if (lon > Math.PI) {
            lon -= (2 * Math.PI);
        }
    
        double Rn = 6378137;
    
        double x = (Rn + alt) * Math.cos(lat) * Math.cos(lon);
        double y = (Rn + alt) * Math.cos(lat) * Math.sin(lon);
        double z = (Rn + alt) * Math.sin(lat);
   
        return new double[] { x, y, z };
    }

**Comment by cogmission** [comment]
@chessweb01 @chetan51 

I'm sorry, I never did see this last comment, and only came to it now because I have some issues surrounding HTM.java's CoordinateEncoder, now.

I refer you to [this link](https://github.com/jswhit/pyproj/blob/master/src/geocent.c) as a reference to the code from which I believe your simplifications came?

The following code doesn't quite look the same upon initial inspection
```
double z = (Rn + alt) * Math.sin(lat);
```
...where "z" is derived from the reference code (line 199 [here](https://github.com/jswhit/pyproj/blob/master/src/geocent.c#L199))

```
*Z = ((Rn * (1 - gi->Geocent_e2)) + Height) * Sin_Lat;
```

My confusion is the lack of reference to the variable "Geocent_e2" which is set up in the init function [here](https://github.com/jswhit/pyproj/blob/master/src/geocent.c#L124) to be:

```
gi->Geocent_a2 = a * a;
gi->Geocent_b2 = b * b;
gi->Geocent_e2 = (gi->Geocent_a2 - gi->Geocent_b2) / gi->Geocent_a2;
```

Somehow I believe, we need to account for the state handling before apply the functions?
